### PR TITLE
sessionctx/stmtctx: fix data race in weak consistency test

### LINKED ISSUE #31965: DATA RACE in the Tracker
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
==================
WARNING: DATA RACE
Write at 0x00c0062e36f0 by goroutine 73:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1009 +0x1e4
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1700 +0x12f
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1708 +0x3c4
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:155 +0x6bc
  github.com/pingcap/tidb/testkit.(*TestKit).MustExec()
      /go/tidb/testkit/testkit.go:77 +0x99
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:140 +0x64c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Previous read at 0x00c0062e36f0 by goroutine 102:
  github.com/pingcap/tidb/util/memory.(*Tracker).Consume()
      /go/tidb/util/memory/tracker.go:318 +0x104
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).sendToRespCh()
      /go/tidb/store/copr/coprocessor.go:550 +0x184
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse()
      /go/tidb/store/copr/coprocessor.go:1016 +0x21f7
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTaskOnce()
      /go/tidb/store/copr/coprocessor.go:789 +0x1b0d
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask()
      /go/tidb/store/copr/coprocessor.go:673 +0x1fc
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).run()
      /go/tidb/store/copr/coprocessor.go:414 +0x184
  github.com/pingcap/tidb/store/copr.(*copIterator).open·dwrap·8()
      /go/tidb/store/copr/coprocessor.go:447 +0x58
Goroutine 73 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  go.uber.org/goleak.VerifyTestMain()
      /home/prow/go/pkg/mod/go.uber.org/goleak@v1.1.12/testmain.go:53 +0x59
  github.com/pingcap/tidb/sessionctx/stmtctx.TestMain()
      /go/tidb/sessionctx/stmtctx/main_test.go:30 +0x224
  main.main()
      _testmain.go:51 +0x264
Goroutine 102 (finished) created at:
  github.com/pingcap/tidb/store/copr.(*copIterator).open()
      /go/tidb/store/copr/coprocessor.go:447 +0xc4
  github.com/pingcap/tidb/store/copr.(*CopClient).Send()
      /go/tidb/store/copr/coprocessor.go:143 +0x10d6
  github.com/pingcap/tidb/distsql.Select()
      /go/tidb/distsql/distsql.go:103 +0x951
  github.com/pingcap/tidb/distsql.SelectWithRuntimeStats()
      /go/tidb/distsql/distsql.go:152 +0xbb
  github.com/pingcap/tidb/executor.selectResultHook.SelectResult()
      /go/tidb/executor/table_reader.go:52 +0x1c5
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).buildResp()
      /go/tidb/executor/table_reader.go:305 +0x4d2
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Open()
      /go/tidb/executor/table_reader.go:199 +0x1358
  github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
      /go/tidb/executor/adapter.go:407 +0x818
  github.com/pingcap/tidb/session.runStmt()
      /go/tidb/session/session.go:1880 +0x70b
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1757 +0xdb3
  github.com/pingcap/tidb/session.(*session).Execute()
      /go/tidb/session/session.go:1448 +0x4b9
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead.func1()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:110 +0x1a1
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:139 +0x62c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Write at 0x00c0062e36f8 by goroutine 73:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1009 +0x1e4
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1700 +0x12f
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1708 +0x3c4
  github.com/pingcap/tidb/testkit.(*TestKit).Exec()
      /go/tidb/testkit/testkit.go:155 +0x6bc
  github.com/pingcap/tidb/testkit.(*TestKit).MustExec()
      /go/tidb/testkit/testkit.go:77 +0x99
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:140 +0x64c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Previous write at 0x00c0062e36f8 by goroutine 102:
  sync/atomic.CompareAndSwapInt64()
      /usr/local/go/src/runtime/race_amd64.s:340 +0xb
  sync/atomic.CompareAndSwapInt64()
      <autogenerated>:1 +0x24
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).sendToRespCh()
      /go/tidb/store/copr/coprocessor.go:550 +0x184
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleCopResponse()
      /go/tidb/store/copr/coprocessor.go:1016 +0x21f7
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTaskOnce()
      /go/tidb/store/copr/coprocessor.go:789 +0x1b0d
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).handleTask()
      /go/tidb/store/copr/coprocessor.go:673 +0x1fc
  github.com/pingcap/tidb/store/copr.(*copIteratorWorker).run()
      /go/tidb/store/copr/coprocessor.go:414 +0x184
  github.com/pingcap/tidb/store/copr.(*copIterator).open·dwrap·8()
      /go/tidb/store/copr/coprocessor.go:447 +0x58
Goroutine 73 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  go.uber.org/goleak.VerifyTestMain()
      /home/prow/go/pkg/mod/go.uber.org/goleak@v1.1.12/testmain.go:53 +0x59
  github.com/pingcap/tidb/sessionctx/stmtctx.TestMain()
      /go/tidb/sessionctx/stmtctx/main_test.go:30 +0x224
  main.main()
      _testmain.go:51 +0x264
Goroutine 102 (finished) created at:
  github.com/pingcap/tidb/store/copr.(*copIterator).open()
      /go/tidb/store/copr/coprocessor.go:447 +0xc4
  github.com/pingcap/tidb/store/copr.(*CopClient).Send()
      /go/tidb/store/copr/coprocessor.go:143 +0x10d6
  github.com/pingcap/tidb/distsql.Select()
      /go/tidb/distsql/distsql.go:103 +0x951
  github.com/pingcap/tidb/distsql.SelectWithRuntimeStats()
      /go/tidb/distsql/distsql.go:152 +0xbb
  github.com/pingcap/tidb/executor.selectResultHook.SelectResult()
      /go/tidb/executor/table_reader.go:52 +0x1c5
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).buildResp()
      /go/tidb/executor/table_reader.go:305 +0x4d2
  github.com/pingcap/tidb/executor.(*TableReaderExecutor).Open()
      /go/tidb/executor/table_reader.go:199 +0x1358
  github.com/pingcap/tidb/executor.(*ExecStmt).Exec()
      /go/tidb/executor/adapter.go:407 +0x818
  github.com/pingcap/tidb/session.runStmt()
      /go/tidb/session/session.go:1880 +0x70b
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1757 +0xdb3
  github.com/pingcap/tidb/session.(*session).Execute()
      /go/tidb/session/session.go:1448 +0x4b9
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead.func1()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:110 +0x1a1
  github.com/pingcap/tidb/sessionctx/stmtctx_test.TestWeakConsistencyRead()
      /go/tidb/sessionctx/stmtctx/stmtctx_test.go:139 +0x62c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
```

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What problem does this PR solve?

Issue Number: close #31965

Problem Summary:

### What is changed and how it works?

Fix a data race in weak consistency test, it's caused by the wrong usage of `Session.Execute`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: sessionctx/stmtctx/stmtctx_test.go [hunk 1]
```diff
@@ -107,7 +107,11 @@ func TestWeakConsistencyRead(t *testing.T) {
 		ctx := context.WithValue(context.Background(), "CheckSelectRequestHook", func(req *kv.Request) {
 			require.Equal(t, req.IsolationLevel, isolationLevel)
 		})
-		tk.Session().Execute(ctx, sql)
+		rss, err := tk.Session().Execute(ctx, sql)
```
Comment (hawkingrei) [inline sessionctx/stmtctx/stmtctx_test.go:110]: use ```tk.Exec``` ?

Comment (you06) [reply sessionctx/stmtctx/stmtctx_test.go:110]: The hook function is passed by context parameter in `Session.Execute`, but `tk.Exec` uses inner context.

**Review by hawkingrei — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**